src/ParallelAlgorithms/Events/ObserveAtExtremum.hpp

```diff
@@ -0,0 +1,404 @@
+    if (not section_observation_key.has_value()) {
+      return std::nullopt;
+    }
+    return {{observers::TypeOfObservation::Reduction,
+             observers::ObservationKey(
+                 subfile_path_ + section_observation_key.value() + ".dat")}};
+  }
+
+  using is_ready_argument_tags = tmpl::list<>;
+
+  template <typename Metavariables, typename ArrayIndex, typename Component>
+  bool is_ready(Parallel::GlobalCache<Metavariables>& /*cache*/,
+                const ArrayIndex& /*array_index*/,
+                const Component* const /*meta*/) const {
+    return true;
+  }
+
+  bool needs_evolved_variables() const override { return true; }
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) override;
+
+ private:
+  std::string subfile_path_;
+  std::string scalar_name_;
+  std::string extremum_type_;
+  std::vector<std::string> additional_tensor_names_{};
+};
+/// @}
+
+/// \cond
+template <typename ObservationValueTag, typename... ObservableTensorTags,
+          typename... NonTensorComputeTags, typename ArraySectionIdTag>
+ObserveAtExtremum<ObservationValueTag, tmpl::list<ObservableTensorTags...>,
+                  tmpl::list<NonTensorComputeTags...>,
+                  ArraySectionIdTag>::ObserveAtExtremum(CkMigrateMessage* msg)
+    : Event(msg) {}
+
+template <typename ObservationValueTag, typename... ObservableTensorTags,
+          typename... NonTensorComputeTags, typename ArraySectionIdTag>
+ObserveAtExtremum<ObservationValueTag, tmpl::list<ObservableTensorTags...>,
+                  tmpl::list<NonTensorComputeTags...>, ArraySectionIdTag>::
+    ObserveAtExtremum(std::string subfile_name,
+                      ObserveTensors observe_tensors)
+    : subfile_path_("/" + subfile_name),
+      scalar_name_(std::move(observe_tensors.scalar_name)),
+      extremum_type_(std::move(observe_tensors.extremum_type)),
+      additional_tensor_names_(std::move(observe_tensors.additional_data)) {}
+
+template <typename ObservationValueTag, typename... ObservableTensorTags,
+          typename... NonTensorComputeTags, typename ArraySectionIdTag>
+ObserveAtExtremum<ObservationValueTag, tmpl::list<ObservableTensorTags...>,
+                  tmpl::list<NonTensorComputeTags...>, ArraySectionIdTag>::
+    ObserveTensors::ObserveTensors(std::string in_scalar,
+                                   std::string in_extremum_type,
+                                   std::vector<std::string> in_additional_data,
+                                   const Options::Context& context)
+    : scalar_name(std::move(in_scalar)),
+      extremum_type(std::move(in_extremum_type)),
+      additional_data(std::move(in_additional_data)) {
+  if (((scalar_name != db::tag_name<ObservableTensorTags>()) and ...)) {
+    PARSE_ERROR(
+        context, "Tensor '"
+                     << scalar_name << "' is not known. Known tensors are: "
+                     << ((db::tag_name<ObservableTensorTags>() + ",") + ...));
```

Comment:
(optional) I think it would be good to check if the tensor is a scalar already here at option parsing, so you don't have to wait for the first observation to get the error (which may not happen until some time into the evolution). If you agree, here's a suggestion:

I think the easiest way to accomplish this is to loop over the `ObservableTensorTags` with a `tmpl::for_each` until you hit the selected tensor, then check `tag::type::rank() == 0`. To deal with optionals you can use `if constexpr (tt::is_a_v<std::optional, tag::type>)`. If it makes things easier I think it would be fine to disallow optionals for the selected scalar (since optionals are used for errors to the analytic solution but you probably want to observe extrema in the density etc), but you can also just get the optional tensor type from `tag::type::value_type`.